### PR TITLE
Fix rex parent_environ case sensitivity on Windows

### LINKED ISSUE #2089: rex environment lookups do not respect platform case-sensitivity rules
On Windows, environment variable keys are case-insensitive for lookup and assignment; Python's wrapper for `os.environ` hides this away behind a `Mapping` interface that normalizes keys to uppercase internally.

However, the `rex` bindings that expose the environment to packages do not handle a provided `parent_environ` dict in a similarly case-insensitive way on Windows. Thus, if a case-sensitive mapping is provided for `parent_environ` (e.g. by copying and modifying `os.environ`), and a package tries to look up an environment variable that is present in `parent_environ` but uses a different casing than `parent_environ`'s entry, it will error similarly to:

```
Error in commands in package '...':
Referenced undefined environment variable: SomeVariable
```

This error does not occur when a `parent_environ` override is not provided, as `rex` will fall back to binding `os.environ`, and thus lookups will be appropriately case-insensitive on Windows.

**Environment**
* Windows 10
* Rez 3.3.0+ (dev build)
* Rez python 3.11.9

**To Reproduce (Windows)**
1. Create and install a package whose `commands` tries to access an environment variable using a TitleCase name, e.g:
  ```python
  name = "env_test"
  build_command = False
  def commands():
      some_value = env.SomeVariable
  ```
2. From a shell, set `SomeVariable` to a value (e.g., from pwsh, `$env:SomeVariable = "1"`).
3. Using `rez-python`, use the `rez` API to create a `ResolvedContext` using the test package and get its environment, passing a `dict` copy of `os.environ` as `parent_environ`:
  ```python
  import os
  from rez.resolved_context import ResolvedContext

  ctx = ResolvedContext(["env_test"], caching=False, add_implicit_packages=False)
  ctx.get_environ(parent_environ=dict(os.environ))
  ```

If you instead omit `parent_environ` or pass an explicit reference to `os.environ`, it will work fine due to the underlying case-insensitive lookup.

### PR DESCRIPTION
**Pull request by thc1006**
Fixes #2089

On Windows, environment variable keys are case-insensitive natively. `os.environ` preserves that contract through Python's `os._Environ` wrapper, but a plain `dict` copy of it (or any user-built dict) does not. Until now, `ActionManager` consumed whatever `Mapping` the caller passed as-is, so a package whose `commands()` referenced `env.SomeVariable` against a `parent_environ` that held the same variable under a different casing would raise:

```
Error in commands in package '...':
Referenced undefined environment variable: SomeVariable
```

Issue #2089 has a clean reproducer that follows exactly this shape: `dict(os.environ)` is passed to `ResolvedContext.get_environ`, and a package looks up `env.SomeVariable` while the dict only has `SOMEVARIABLE`.

The fix wraps a caller-supplied `parent_environ` in a small read-only, case-normalized snapshot (`_CaseInsensitiveEnvironProxy`), gated on `platform_.name == "windows"`. `os.environ` and `None` are used as-is (they are already case-insensitive on Windows), so their identity and liveness are preserved. On Linux and macOS a caller-supplied mapping retains the existing identity passthrough. Keys are upper-cased once at construction, i.e. a point-in-time snapshot, which matches how `parent_environ` is fixed for an executor's lifetime; iteration matches what `os.environ` yields on Windows, and when the input contains keys that differ only by case the last one encountered wins, mirroring how `os.environ` collapses duplicates. The proxy also exposes `.copy()` (returning a plain dict with upper-cased keys) for callers that expect a dict-like `parent_environ`. The `not isinstance(...)` guard in the gate keeps wrapping idempotent.

Tests:
- `test_case_insensitive_environ_proxy` unit-tests the proxy directly and runs on every platform, so the core normalization logic has real CI coverage: get/contains across casings, missing-key `KeyError`, `.copy()`, iteration/length, and last-write-wins collisions.
- `test_parent_environ_case_insensitive_on_windows` (Windows-only via `@unittest.skipIf`) drives the real package-commands entrypoint — `env.<MixedCase>` — through both `execute_function` and `execute_code` via the `_test` helper, alongside direct `parent_environ[...]` lookups across casings.
- `test_parent_environ_identity_passthrough_on_non_windows` (non-Windows) asserts `ex.manager.parent_environ is env` to guard against accidental wrapping.

Verified locally on Linux: `python -m unittest rez.tests.test_rex` → 17 passed + 1 Windows-only skip. The full `rez-selftest`, `flake8`, and `mypy` matrix runs in CI.

Scope: this makes lookups against a caller-supplied `parent_environ` case-insensitive on Windows — the exact #2089 reproducer. Variables set *within* rex code (`ActionManager.environ`) are stored verbatim and are not case-folded; that is a separate, pre-existing gap and intentionally out of scope here — tracked in #2164.

Reported by @nrusch. Independent diagnosis previously posted in #2093 (closed because the author could not satisfy the EasyCLA/DCO requirement, not for technical reasons).

Disclosure: Claude Code (Opus 4.8) used for adversarial review, edge-case analysis, test design, and implementation of the review feedback.

### CODE REVIEW

**Comment by thc1006** [comment]
Closing+reopening to retrigger Read the Docs (initial build failed during clone, 2s duration, generic clone error — fork branch is public and accessible).

**Comment by thc1006** [comment]
Status note for reviewers:

- The six core GitHub Actions workflows (`tests`, `flake8`, `mypy`, `copyright`, `installation`, `benchmark`) are queued in `action_required` state, awaiting maintainer approval per the first-time-contributor policy. Nothing to do on my side.
- The Read the Docs failure is `git fetch origin pull/2098/head` returning `fatal: couldn't find remote ref pull/2098/head`. The GitHub git API for this PR currently exposes only `refs/pull/2098/merge`; `refs/pull/2098/head` is missing from `matching-refs/pull/2098`, while other open PRs (e.g. #2090, #2091) have both refs as expected. An amend force-push was attempted to nudge GitHub into recreating the head ref but it has not reappeared. This appears to be a GitHub-side propagation glitch independent of the diff. Approving the workflows may incidentally fix it; if not, an RTD admin rebuild should clear it.
- DCO and EasyCLA are both green.

Local pre-push validation: `pytest src/rez/tests/test_rex.py` is green (15 pass + 1 Windows-only skip); full `rez-selftest` on a hermetic Python 3.11.14 container reports `Ran 259 tests in 24.997s, OK (skipped=10)`; `flake8 src/rez/rex.py src/rez/tests/test_rex.py` is clean; `mypy | mypy-baseline filter --ignore-categories=note --allow-unsynced` reports no new violations above the pinned baseline.

File: src/rez/tests/test_rex.py [hunk 1]
```diff
@@ -547,6 +547,39 @@ def test_intersects_ephemerals(self) -> None:
...
+        """Regression for #2089.
+
+        On Windows env vars are case-insensitive; os.environ already
+        behaves that way, but a plain dict copy of it doesn't. Passing
+        such a dict as parent_environ used to break env.<MixedCase>
+        lookups inside commands(). We now wrap on the Windows side.
+        """
+        from rez.utils.platform_ import platform_
+        if platform_.name != "windows":
+            self.skipTest("Windows-only behaviour")
```
Comment (maxnbk) [inline src/rez/tests/test_rex.py]: Let's use the @unittest.skipIf decorator instead of the in-test skip, to prevent the test from even firing on non-windows hosts.

File: src/rez/rex.py [hunk 2]
```diff
@@ -447,6 +457,35 @@ def _keytoken(self, key):
...
+        # keys are upper-cased once on the way in; same shape as
+        # os.environ on Windows. Last write wins on case collisions,
+        # which mirrors how os.environ resolves them too.
+        self._data = {k.upper(): v for k, v in data.items()}
+
+    def __getitem__(self, key):
+        return self._data[key.upper()]
+
+    def __contains__(self, key):
+        return isinstance(key, str) and key.upper() in self._data
```
Comment (maxnbk) [inline src/rez/rex.py]: I don't believe the `isinstance(key, str)` check is necessary even if harmless. All callers use string keys. The guard means that `non_string_key in proxy` returns `False` instead of raising `TypeError` (which is what a normal dict does).

**Review by maxnbk — CHANGES_REQUESTED** [review]
Apologies for taking a long time to get to this review.

I've commented on a few specific items.

More generally, `_CaseInsensitiveEnvironProxy` could have a `.copy()` method added, which would be defensive against future code calling `self.parent_environ.copy()` to return a plain `dict` with uppercased keys.

`ActionManager.__init__` docstring could document the Windows wrapping behavior.

And last, I feel it would be useful to consider a non-Windows identity passthrough test. i.e. assert `ex.manager.parent_environ is env` on non-Windows to guard against accidental wrapping. With that in mind, the test skip could be avoided and just assert differently for different host types, or be a different test skipping windows hosts, either way.

If you are using an LLM to produce this PR, please add a line to the PR description disclosing the usage, i.e. "Disclosure: <model> used for <task> (whichever things it was used for.)

File: src/rez/tests/test_rex.py (hunk 1 above)
Comment (thc1006) [reply src/rez/tests/test_rex.py]: Done in 1c60859e — switched to `@unittest.skipIf(platform_.name != "windows", ...)` with `platform_` imported at module level. I also added `test_parent_environ_identity_passthrough_on_non_windows` (asserts `parent_environ is env` on non-Windows) and a platform-independent `test_case_insensitive_environ_proxy`, so the core normalization logic is exercised on Linux/macOS CI rather than skipped everywhere but Windows.

File: src/rez/rex.py (hunk 2 above)
Comment (thc1006) [reply src/rez/rex.py]: Done in 1c60859e — dropped the guard; `__contains__` is now just `return key.upper() in self._data`. Agreed the type contract is `Mapping[str, str]`, so non-str keys are out of contract anyway.

**Review by maxnbk — CHANGES_REQUESTED** [review]
Tests failed, I believe it is because rex `env` is shadowing your variables. I am guessing something like this might fix it. GitHub UI is misbehaving so I wasn't able to apply it as a suggestion to click apply for:
```
         """Regression for #2089.
         ...
         """
-        env = {"SomeVariable": "covfefe"}
-        ex = self._create_executor(env)
+        parent_env = {"SomeVariable": "covfefe"}
+        ex = self._create_executor(parent_env)

         # any casing of the key resolves
         self.assertEqual(ex.manager.parent_environ["SomeVariable"], "covfefe")
         ...

         # end-to-end via the real package-commands entrypoint
         def _rex() -> None:
             env.RESULT = env.SOMEVARIABLE  # noqa: F821 - rex builtin

         self._test(
             func=_rex,
-            env={"SomeVariable": "covfefe"},
+            env=parent_env,
             expected_actions=[Setenv("RESULT", "covfefe")],
             expected_output={"RESULT": "covfefe"},
         )
```
(It may be wise to check the tests for any similar pattern and update those just on the off-chance someone introduces rex in any of those tests in the future and the same gotcha resurfaces too)

**Comment by thc1006** [comment]
Thanks for the diagnosis — that was exactly it. `execute_function` rebuilds the function with `closure=func.__closure__`, so once the test declared a local `env` dict the nested `_rex` captured it as a closure cell (`LOAD_DEREF`) and shadowed the rex `env` builtin injected into globals — `env.SOMEVARIABLE` then hit a plain `dict`. `execute_code` was fine (fresh compile from source), which is why only the function path failed.

Fixed in 1c3299ea: renamed the caller dict to `parent_env` in both new tests, so `_rex`'s `env` is a plain global again and resolves to the rex builtin. Added a short comment on the test flagging the gotcha.

Per your future-proofing note, I audited the rest of `test_rex.py`: no other test declares a local `env` — they all pass `env=` as a kwarg to `_test`, so none can shadow the builtin.

I also closed the codecov gap on the Windows-only branch (it's skipped on the Linux/macOS jobs):
- `test_parent_environ_os_environ_passthrough` — covers the `parent_environ is os.environ` identity branch.
- `test_parent_environ_case_insensitive_windows_mocked` — mocks `platform_.name` so the case-insensitive wrapping branch and the full `env.<MixedCase>` end-to-end run on every host. As a bonus that validates the Windows logic off-Windows, which is what would have caught this shadowing bug locally.

**Review by maxnbk — APPROVED** [review]
It LGTM now. Thank you for making the requested changes @thc1006 . I marked Approve (from me) and put this into "Next" milestone for merge in the next release. No further action needed unless some other reviewer requests changes.

File: src/rez/rex.py [hunk 3]
```diff
@@ -447,6 +465,47 @@ def _keytoken(self, key):
         return self.interpreter.get_key_token(key)
 
 
+class _CaseInsensitiveEnvironProxy(Mapping):
+    """Read-only, case-insensitive snapshot of an env-var mapping.
+
+    Used by `ActionManager` to wrap a caller-supplied `parent_environ`
```
Comment (JeanChristopheMorinPerso) [inline src/rez/rex.py]:
```suggestion
    Used by :class:`ActionManager` to wrap a caller-supplied ``parent_environ``
```

File: src/rez/rex.py [hunk 4]
```diff
@@ -187,7 +187,10 @@ def __init__(self, interpreter: ActionInterpreter,
         interpreter: string or `ActionInterpreter`
             the interpreter to use when executing rex actions
         parent_environ: environment to execute the actions within. If None,
-            defaults to the current environment.
+            defaults to the current environment. On Windows, a caller-supplied
+            mapping is wrapped in a read-only, case-insensitive snapshot so that
+            lookups against it are case-insensitive, matching how ``os.environ``
+            behaves natively (see #2089); `os.environ` and None are used as-is.
```
Comment (JeanChristopheMorinPerso) [inline src/rez/rex.py]:
```suggestion
            lookups against it are case-insensitive, matching how :data:`os.environ`
            behaves natively (see #2089); :data:`os.environ` and None are used as-is.
```

File: src/rez/rex.py [hunk 5]
```diff
@@ -447,6 +465,47 @@ def _keytoken(self, key):
         return self.interpreter.get_key_token(key)
 
 
+class _CaseInsensitiveEnvironProxy(Mapping):
+    """Read-only, case-insensitive snapshot of an env-var mapping.
+
+    Used by `ActionManager` to wrap a caller-supplied `parent_environ`
+    on Windows so lookups against it are case-insensitive regardless of
+    the casing used to populate the input dict, matching how os.environ
```
Comment (JeanChristopheMorinPerso) [inline src/rez/rex.py]:
```suggestion
    the casing used to populate the input dict, matching how :data:`os.environ`
```

File: src/rez/rex.py [hunk 6]
```diff
@@ -447,6 +465,47 @@ def _keytoken(self, key):
...
+    behaves natively on Windows.
+
+    Keys are normalized (upper-cased) once at construction, so this is a
+    point-in-time snapshot: later mutations to the source mapping are not
+    reflected. That matches how `parent_environ` is used in practice (fixed
```
Comment (JeanChristopheMorinPerso) [inline src/rez/rex.py]:
```suggestion
    reflected. That matches how ``parent_environ`` is used in practice (fixed
```

File: src/rez/rex.py [hunk 7]
```diff
@@ -447,6 +465,47 @@ def _keytoken(self, key):
         return self.interpreter.get_key_token(key)
 
 
+class _CaseInsensitiveEnvironProxy(Mapping):
```
Comment (JeanChristopheMorinPerso) [inline src/rez/rex.py:468]:
```suggestion
class _CaseInsensitiveEnvironProxy(Mapping[str, str]):
```
?

Comment (thc1006) [reply src/rez/rex.py:468]: tl;dr: happy to take the `Mapping[str, str]` change, but it should use `typing.Mapping` to stay 3.8-safe. Separately, the failing Ruff check looks like a pre-existing issue on `main`, not something in this PR.

On the base class: `Mapping` here is imported from `collections.abc`, and using one of those as a subscripted base (`class X(Mapping[str, str])`) only works on Python 3.9+. On 3.8 it raises `TypeError` at import, and we still support 3.8 (`python_requires >= 3.8`). The other parametrized bases in the repo, like `AttrDictWrapper(MutableMapping[str, Any])` and `PackageOrderList(List[...])`, go through `typing` for that reason. I can switch this to `from typing import Mapping` and apply `Mapping[str, str]`, or leave it bare, whichever you prefer.

On the Ruff failure: it's `ISC004` at `src/rez/utils/platform_.py:524` (the PowerShell command list), which this branch doesn't touch. It reproduces on a clean `main` (`4946f9de`) with ruff 0.16.0, the newest release matching `>= 0.15.0` that `ruff-action` pulls in. So it reads as the new ISC004 rule catching an existing line rather than anything from here. I'm happy to fold the one-line fix into this PR if that unblocks things, otherwise it may fit better as its own change on `main`.

Thanks for the sphinx pass too, the `:data:`/`:class:` refs read a lot cleaner.

**Comment by maxnbk** [comment]
The ruff issue is fixed on main, so a rebase will pick it up and validate that no new ruff issues pop up.
